src/TraceEvent/EventPipe/EventCache.cs

```diff
@@ -0,0 +1,282 @@
+                    OnEvent?.Invoke(ref eventMarker.Header);
+                }
+                else
+                {
+                    thread.Events.Enqueue(eventMarker);
+
+                }
+
+                cursor += eventMarker.Header.TotalNonHeaderSize + eventMarker.Header.HeaderSize;
+                EventMarker lastEvent = eventMarker;
+                eventMarker = new EventMarker(buffer);
+                eventMarker.Header = lastEvent.Header;
+            }
+            thread.LastCachedEventTimestamp = timestamp;
+        }
+
+        public unsafe void ProcessSequencePointBlock(byte[] sequencePointBytes)
+        {
+            if(sequencePointBytes.Length < 12)
+            {
+                Debug.Assert(false, "Bad sequence point block length");
+                return;
+            }
+            long timestamp = BitConverter.ToInt64(sequencePointBytes, 0);
+            int threadCount = BitConverter.ToInt32(sequencePointBytes, 8);
+            if(sequencePointBytes.Length < 12 + threadCount*12)
+            {
+                Debug.Assert(false, "Bad sequence point block length");
+                return;
+            }
+            SortAndDispatch(timestamp);
+            foreach(EventCacheThread thread in _threads.Values)
+            {
+                Debug.Assert(thread.Events.Count == 0, "There shouldn't be any pending events after a sequence point");
+                thread.Events.Clear();
+                thread.Events.TrimExcess();
+            }
+
+            int cursor = 12;
+            for(int i = 0; i < threadCount; i++)
+            {
+                long captureThreadId = BitConverter.ToInt64(sequencePointBytes, cursor);
+                int sequenceNumber = BitConverter.ToInt32(sequencePointBytes, cursor + 8);
+                if (!_threads.TryGetValue(captureThreadId, out EventCacheThread thread))
+                {
+                    if(sequenceNumber > 0)
+                    {
+                        OnEventsDropped?.Invoke(sequenceNumber);
+                    }
+                    thread = new EventCacheThread();
+                    thread.SequenceNumber = sequenceNumber;
+                    AddThread(captureThreadId, thread);
+                }
+                else
+                {
+                    int droppedEvents = unchecked(sequenceNumber - thread.SequenceNumber);
+                    if (droppedEvents > 0)
+                    {
+                        OnEventsDropped?.Invoke(droppedEvents);
+                    }
+                    else
+                    {
+                        // When a thread id is recycled the sequenceNumber can abruptly reset to 1 which
+                        // makes droppedEvents go negative
+                        Debug.Assert(droppedEvents == 0 || sequenceNumber == 1);
+                    }
+                    thread.SequenceNumber = sequenceNumber;
+                }
+                cursor += 12;
```

Comment:
Lots of 12s around here.  It might be good to turn this into a constant so that it's clear what the 12s mean.  I'm assuming its the size of the sequence point block?

Reply:
There are actually two things which are both 12. The sequence point block has a 12 byte header:
8 byte Timestamp + 4 byte count of threads

And a 12 byte per-thread entry:
8 byte thread id + 4 byte sequence number

I added some const ints to better describe the magic numbers.

Reply:
Awesome.  Thanks.